saffron/src/read_proof.rs

```diff
@@ -0,0 +1,381 @@
+
+    let coms_and_evaluations = vec![
+        Evaluation {
+            commitment: PolyComm {
+                chunks: vec![*data_comm],
+            },
+            evaluations: vec![vec![proof.data_eval]],
+        },
+        Evaluation {
+            commitment: PolyComm {
+                chunks: vec![proof.query_comm],
+            },
+            evaluations: vec![vec![proof.query_eval]],
+        },
+        Evaluation {
+            commitment: PolyComm {
+                chunks: vec![proof.answer_comm],
+            },
+            evaluations: vec![vec![proof.answer_eval]],
+        },
+        Evaluation {
+            commitment: PolyComm {
+                chunks: vec![proof.quotient_comm],
+            },
+            evaluations: vec![vec![quotient_eval]],
+        },
+    ];
+
+    let combined_inner_product = {
+        let evaluations: Vec<_> = coms_and_evaluations
+            .iter()
+            .map(|Evaluation { evaluations, .. }| evaluations.clone())
+            .collect();
+
+        combined_inner_product(&polyscale, &evalscale, evaluations.as_slice())
+    };
+
+    srs.verify(
+        group_map,
+        &mut [BatchEvaluationProof {
+            sponge: fq_sponge,
+            evaluation_points: vec![evaluation_point],
+            polyscale,
+            evalscale,
+            evaluations: coms_and_evaluations,
+            opening: &proof.opening_proof,
+            combined_inner_product,
+        }],
+        rng,
+    )
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{prove, verify, ReadProof};
+    use crate::{env, Curve, ScalarField, SRS_SIZE};
+    use ark_ec::AffineRepr;
+    use ark_ff::{One, UniformRand};
+    use ark_poly::{univariate::DensePolynomial, Evaluations};
+    use kimchi::{circuits::domains::EvaluationDomains, groupmap::GroupMap};
+    use mina_curves::pasta::{Fp, Vesta};
+    use once_cell::sync::Lazy;
+    use poly_commitment::{commitment::CommitmentCurve, ipa::SRS, SRS as _};
+    use proptest::prelude::*;
+
+    static SRS: Lazy<SRS<Vesta>> = Lazy::new(|| {
+        if let Ok(srs) = std::env::var("SRS_FILEPATH") {
+            env::get_srs_from_cache(srs)
+        } else {
+            SRS::create(SRS_SIZE)
+        }
+    });
+
+    static DOMAIN: Lazy<EvaluationDomains<ScalarField>> =
+        Lazy::new(|| EvaluationDomains::<ScalarField>::create(SRS_SIZE).unwrap());
+
+    static GROUP_MAP: Lazy<<Vesta as CommitmentCurve>::Map> =
+        Lazy::new(<Vesta as CommitmentCurve>::Map::setup);
+
+    #[test]
+    fn test_read_proof_completeness_soundness() {
+        let mut rng = o1_utils::tests::make_test_rng(None);
+
+        let data: Vec<ScalarField> = {
+            let mut data = vec![];
+            (0..SRS_SIZE)
+                .into_iter()
+                .for_each(|_| data.push(Fp::rand(&mut rng)));
+            data
+        };
+
+        let data_poly: DensePolynomial<ScalarField> =
+            Evaluations::from_vec_and_domain(data.clone(), (*DOMAIN).d1).interpolate();
```

Comment:
Nit: you can avoid cloning here.

Reply:
here I can't it seems because `from_vec_and_domain` consumes the data